File: exercism/haskell/complex-numbers/src/ComplexNumbers.hs
module ComplexNumbers
  ( Complex
  , conjugate
  , abs
  , exp
  , real
  , imaginary
  , mul
  , add
  , sub
  , div
  , complex
  )
where

import Prelude hiding (abs, div, exp)
import qualified Prelude

data Complex a = a :+ a deriving (Eq, Show)

complex :: (a, a) -> Complex a
complex = uncurry (:+)

conjugate :: Num a => Complex a -> Complex a
conjugate (re :+ im) = re :+ (- im)

abs :: Floating a => Complex a -> a
abs (re :+ im) = sqrt $ re * re + im * im

real :: Num a => Complex a -> a
real (re :+ _) = re

imaginary :: Num a => Complex a -> a
imaginary (_ :+ im) = im

exp :: Floating a => Complex a -> Complex a
exp (a :+ b) = (expA * cos b) :+ (expA * sin b)
  where
    expA = Prelude.exp a

mul :: Num a => Complex a -> Complex a -> Complex a
mul (a :+ b) (c :+ d) = (a * c - b * d) :+ (b * c + a * d)

add :: Num a => Complex a -> Complex a -> Complex a
add (a :+ b) (c :+ d) = (a + c) :+ (b + d)

sub :: Num a => Complex a -> Complex a -> Complex a
sub (a :+ b) (c :+ d) = (a - c) :+ (b - d)

div :: Fractional a => Complex a -> Complex a -> Complex a
div (a :+ b) (c :+ d) =
  ((a * c + b * d) / s) :+ ((b * c - a * d) / s)
  where
    s = c * c + d * d

File: exercism/kotlin/complex-numbers/src/main/kotlin/ComplexNumbers.kt
import kotlin.math.cos
import kotlin.math.exp
import kotlin.math.sin
import kotlin.math.sqrt

data class ComplexNumber(val real: Double = 0.0, val imag: Double = 0.0) {
    val abs = sqrt(real * real + imag * imag)

    operator fun plus(that: ComplexNumber) =
        ComplexNumber(real + that.real, imag + that.imag)

    operator fun minus(that: ComplexNumber) =
        ComplexNumber(real - that.real, imag - that.imag)

    operator fun times(that: ComplexNumber) =
        ComplexNumber(
            real * that.real - imag * that.imag,
            imag * that.real + real * that.imag
        )

    operator fun div(that: ComplexNumber) = (that.real * that.real + that.imag * that.imag).let { s ->
        ComplexNumber(
            (real * that.real + imag * that.imag) / s,
            (imag * that.real - real * that.imag) / s
        )
    }

    fun conjugate() = ComplexNumber(real, -imag)
}

fun exponential(c: ComplexNumber): ComplexNumber = exp(c.real).let { expA ->
    ComplexNumber(expA * cos(c.imag), expA * sin(c.imag))
}